tor/controller.go

```diff
@@ -338,60 +338,79 @@ func (c *Controller) ProtocolInfo() ([]string, string, string, error) {
 	return authMethods, cookieFilePath, torVersion, nil
 }
 
-// VirtToTargPorts is a mapping of virtual ports to target ports. When creating
-// an onion service, it will be listening externally on each virtual port. Each
-// virtual port can then be mapped to one or many target ports internally. When
-// accessing the onion service at a specific virtual port, it will forward the
-// traffic to a mapped randomly chosen target port.
-type VirtToTargPorts = map[int]map[int]struct{}
-
-// AddOnionV2 creates a new v2 onion service and returns its onion address(es).
-// Once created, the new onion service will remain active until the connection
-// between the controller and the Tor server is closed. The path to a private
-// key can be provided in order to restore a previously created onion service.
-// If a file at this path does not exist, a new onion service will be created
-// and its private key will be saved to a file at this path. A mapping of
-// virtual ports to target ports should also be provided. Each virtual port will
-// be the ports where the onion service can be reached at, while the mapped
-// target ports will be the ports where the onion service is running locally.
-func (c *Controller) AddOnionV2(privateKeyFilename string,
-	virtToTargPorts VirtToTargPorts) ([]*OnionAddr, error) {
+// OnionType denotes the type of the onion service.
+type OnionType int
 
+const (
+	// V2 denotes that the onion service is V2.
+	V2 OnionType = iota
+
+	// V3 denotes that the onion service is V3.
+	V3
+)
+
+// AddOnionConfig houses all of the required paramaters in order to succesfully
+// create a new onion service or restore an existing one.
+type AddOnionConfig struct {
+	// Type denotes the type of the onion service that should be created.
+	Type OnionType
+
+	// VirtualPort is the externally reachable port of the onion address.
+	VirtualPort int
+
+	// TargetPorts is the set of ports that the service will be listening on
+	// locally. The virutal port will forward the traffic to one randomly
```

Comment:
where is this randomly chosen?

Reply:
By the Tor server itself. Clarified this in the comments.